subsys/modem/modem_stats.c

```diff
@@ -0,0 +1,165 @@
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(modem_stats);
+
+static struct k_spinlock stats_buffer_lock;
+static sys_slist_t stats_buffer_list;
+
+static struct modem_stats_buffer *stats_buffer_from_node(sys_snode_t *node)
+{
+	return (struct modem_stats_buffer *)node;
+}
+
+static void stats_buffer_list_append(struct modem_stats_buffer *buffer)
+{
+	K_SPINLOCK(&stats_buffer_lock) {
+		sys_slist_append(&stats_buffer_list, &buffer->node);
+	}
+}
+
+static struct modem_stats_buffer *stats_buffer_list_first(void)
+{
+	struct modem_stats_buffer *first;
+
+	K_SPINLOCK(&stats_buffer_lock) {
+		first = stats_buffer_from_node(sys_slist_peek_head(&stats_buffer_list));
+	}
+
+	return first;
+}
+
+static struct modem_stats_buffer *stats_buffer_list_next(struct modem_stats_buffer *buffer)
+{
+	struct modem_stats_buffer *next;
+
+	K_SPINLOCK(&stats_buffer_lock) {
+		next = stats_buffer_from_node(sys_slist_peek_next(&buffer->node));
+	}
+
+	return next;
+}
+
+static uint8_t percent_used(uint32_t max, uint32_t cap)
+{
+	uint64_t percent;
+
+	if (max == 0) {
+		return 0;
+	}
+
+	if (max == cap) {
+		return 100;
+	}
+
+	percent = 100;
+	percent <<= 16;
+	percent /= cap;
+	percent *= max;
+	percent >>= 16;
+
+	return (uint8_t)percent;
+}
+
+static void stats_buffer_get_and_clear_max(struct modem_stats_buffer *buffer, uint32_t *max)
+{
+	K_SPINLOCK(&stats_buffer_lock) {
+		*max = buffer->max;
+		buffer->max = 0;
+	}
+}
+
+static bool stats_buffer_size_is_valid(const struct modem_stats_buffer *buffer, uint32_t size)
+{
+	return size <= buffer->capacity;
+}
+
+static void stats_buffer_log_invalid_size(const struct modem_stats_buffer *buffer, uint32_t size)
+{
+	LOG_ERR("%s: size (%u) exceeds capacity (%u)", buffer->name, size, buffer->capacity);
+}
+
+static void stats_buffer_update_max(struct modem_stats_buffer *buffer, uint32_t size)
+{
+	K_SPINLOCK(&stats_buffer_lock) {
+		if (buffer->max < size) {
+			buffer->max = size;
+		}
+	}
+}
+
+static void stats_buffer_print_to_shell(const struct shell *sh,
+					const struct modem_stats_buffer *buffer, uint32_t max)
+{
+	shell_print(sh, "%s: used at most: %u of %u (%u%%)", buffer->name, max,
+		    buffer->capacity, percent_used(max, buffer->capacity));
+}
+
+static int stats_buffer_shell_cmd_handler(const struct shell *sh, size_t argc, char **argv)
+{
+	struct modem_stats_buffer *buffer;
+	uint32_t max;
+
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+
+	buffer = stats_buffer_list_first();
+
+	if (buffer == NULL) {
+		shell_print(sh, "no buffers exist");
+		return 0;
+	}
+
+	while (buffer != NULL) {
+		stats_buffer_get_and_clear_max(buffer, &max);
+		stats_buffer_print_to_shell(sh, buffer, max);
+		buffer = stats_buffer_list_next(buffer);
+	}
+
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	sub_stats_cmds,
+	SHELL_CMD(buffer, NULL, "Get buffer statistics", stats_buffer_shell_cmd_handler),
+	SHELL_SUBCMD_SET_END
+);
+
+SHELL_CMD_REGISTER(modem_stats, &sub_stats_cmds, "Modem statistics commands", NULL);
```

Comment:
What is actually the shell command to issue? (I haven't yet used Zephyr's shell.)

Reply:
`modem_stats buffer`